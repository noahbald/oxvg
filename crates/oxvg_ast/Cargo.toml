[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
