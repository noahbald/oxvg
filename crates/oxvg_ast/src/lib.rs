#[macro_use]
extern crate bitflags;

#[cfg(feature = "style")]
#[macro_use]
extern crate smallvec;

pub mod atom;
pub mod attribute;
pub mod class_list;
pub mod document;
pub mod element;
pub mod implementations;
pub mod name;
pub mod node;

#[cfg(feature = "visitor")]
pub mod visitor;

#[cfg(feature = "parse")]
pub mod parse;

#[cfg(feature = "serialize")]
pub mod serialize;

#[cfg(feature = "selectors")]
pub mod selectors;

#[cfg(feature = "style")]
pub mod style;
