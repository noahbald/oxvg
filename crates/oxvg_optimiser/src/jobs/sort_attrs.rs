use oxvg_ast::{
    attribute::Attributes,
    element::Element,
    visitor::{Context, Visitor},
};
use serde::{Deserialize, Serialize};

#[cfg(feature = "wasm")]
use tsify::Tsify;

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "napi", napi)]
#[derive(Deserialize, Serialize, Debug, Clone, Default, PartialEq)]
#[serde(rename_all = "lowercase")]
/// The method for ordering xmlns attributes
pub enum XMLNSOrder {
    /// Sort xmlns attributes alphabetically
    Alphabetical,
    #[default]
    /// Keep xmlns attributes at the front of the list
    Front,
}

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "napi", napi(object))]
#[derive(Deserialize, Serialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
/// Sorts attributes into a predictable order.
///
/// This doesn't affect the size of a document but will likely improve readability
/// and compression of the document.
///
/// # Correctness
///
/// This job should never visually change the document.
///
/// # Errors
///
/// Never.
///
/// If this job produces an error or panic, please raise an [issue](https://github.com/noahbald/oxvg/issues)
pub struct SortAttrs {
    /// A list of attributes in a given order.
    #[cfg_attr(feature = "wasm", tsify(optional))]
    pub order: Option<Vec<String>>,
    /// The method for ordering xmlns attributes
    #[cfg_attr(feature = "wasm", tsify(optional))]
    pub xmlns_order: Option<XMLNSOrder>,
}

impl<'arena, E: Element<'arena>> Visitor<'arena, E> for SortAttrs {
    type Error = String;

    fn element(
        &self,
        element: &mut E,
        _context: &mut Context<'arena, '_, '_, E>,
    ) -> Result<(), String> {
        let order = self.order.as_ref().unwrap_or_else(|| &DEFAULT_ORDER);
        let xmlns_order = self.xmlns_order.is_none() || self.xmlns_order == Some(XMLNSOrder::Front);
        element.attributes().sort(order, xmlns_order);

        Ok(())
    }
}

lazy_static! {
    pub static ref DEFAULT_ORDER: Vec<String> = vec![
        String::from("id"),
        String::from("width"),
        String::from("height"),
        String::from("x"),
        String::from("x1"),
        String::from("x2"),
        String::from("y"),
        String::from("y1"),
        String::from("y2"),
        String::from("cx"),
        String::from("cy"),
        String::from("r"),
        String::from("fill"),
        String::from("stroke"),
        String::from("marker"),
        String::from("d"),
        String::from("points"),
    ];
}

#[test]
fn sort_attrs() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "sortAttrs": {} }"#,
        Some(
            r#"<svg r="" b="" x2="" cx="" y1="" a="" y="" y2="" x1="" cy="" x="">
    <!-- sort according to default list alphabetically -->
</svg>"#
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "sortAttrs": {} }"#,
        Some(
            r#"<svg a="" fill-opacity="" stroke="" fill="" stroke-opacity="">
    <!-- sort derived attributes like fill and fill-opacity -->
</svg>"#
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "sortAttrs": {} }"#,
        Some(
            r#"<svg xmlns:editor2="link2" fill="" b="" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:editor1="link1" xmlns="" d="">
    <!-- put xmlns and namespace attributes before others by default -->
    <rect editor2:b="" editor1:b="" editor2:a="" editor1:a="" />
</svg>"#
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "sortAttrs": { "xmlnsOrder": "alphabetical" } }"#,
        Some(
            r#"<svg foo="bar" xmlns="http://www.w3.org/2000/svg" height="10" baz="quux" width="10" hello="world">
    <!-- optionally sort xmlns attributes alphabetically -->
    <rect x="0" y="0" width="100" height="100" stroke-width="1" stroke-linejoin="round" fill="red" stroke="orange" xmlns="http://www.w3.org/2000/svg"/>
    test
</svg>"#
        ),
    )?);

    Ok(())
}
