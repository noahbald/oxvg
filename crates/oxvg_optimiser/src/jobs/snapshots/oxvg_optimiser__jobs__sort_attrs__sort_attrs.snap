---
source: crates/oxvg_optimiser/src/jobs/sort_attrs.rs
expression: "test_config(r#\"{ \"sortAttrs\": {} }\"#,\nSome(r#\"<svg r=\"\" b=\"\" x2=\"\" cx=\"\" y1=\"\" a=\"\" y=\"\" y2=\"\" x1=\"\" cy=\"\" x=\"\">\n    <!-- sort according to default list alphabetically -->\n</svg>\"#),)?"
---
<svg x="" x1="" x2="" y="" y1="" y2="" cx="" cy="" r="" a="" b="">
    <!-- sort according to default list alphabetically -->
</svg>
